### PR TITLE
[DON'T MERGE] Prototype on how to possibly store parsed ANSI

### PR DESCRIPTION
**Pull request by lgeiger**
This is intended to outline a fairly easy way to improve ANSI rendering performance without implementing a terminal like described [here](https://github.com/nteract/meeting-minutes/blob/master/2017/05-22.md#ansi-handling-in-nteract--hydrogen). The advantage over a terminal is mainly that it's easier to implement since we only need to copy paste existing code 😄 .

This would be the next steps:
- Add `@nteract/ansi-parse` which exports [`ansiToInlineStyle`](https://github.com/nteract/ansi-to-react/blob/master/src/index.js#L7-L40)
- Add new transform using [`inlineBundleToReact`](https://github.com/nteract/ansi-to-react/blob/master/src/index.js#L65-L70) to render the json from `@nteract/ansi-parse`. This would accept a new mime type like `application/ansi+json` or `application/nteract.ansi+json`.
- Update [`reduceOutput`](https://github.com/nteract/nteract/blob/master/src/notebook/reducers/document.js#L61) to always use `createImmutableOutput`.

### CODE REVIEW

File: packages/commutable/src/v4.js [hunk 1]
```diff
@@ -160,6 +161,12 @@ export function cleanMimeAtKey(
   previous: ImmutableMimeBundle,
   key: string
 ): ImmutableMimeBundle {
+  if (key === "text/plain") {
+    previous = previous.set(
+      "application/ansi+json",
```
Comment (rgbkrk) [inline packages/commutable/src/v4.js:166]: This is only intended for at-runtime right? We still want to end up with `text/plain` for the notebook format.

Also... text/plain won't get additional text. New display_data are wholly distinct.

File: packages/commutable/src/v4.js [hunk 2]
```diff
@@ -224,7 +231,8 @@ export function createImmutableOutput(output: Output): ImmutableOutput {
       return Immutable.Map({
         output_type: output.output_type,
         name: output.name,
-        text: demultiline(output.text)
+        text: demultiline(output.text),
+        parsed_ansi: parse(demultiline(output.text))
```
Comment (rgbkrk) [inline packages/commutable/src/v4.js:235]: Smart.

File: packages/commutable/src/v4.js [hunk 3]
```diff
@@ -233,7 +241,12 @@ export function createImmutableOutput(output: Output): ImmutableOutput {
         evalue: output.evalue,
         // Note: this is one of the cases where the Array of strings (for traceback)
         // is part of the format, not a multiline string
-        traceback: Immutable.List(output.traceback)
+        traceback: Immutable.List(output.traceback),
+        parsed_ansi: parse(
+          output.traceback
+            ? output.traceback.join("\n")
+            : `${output.ename}: ${output.evalue}`
+        )
```
Comment (rgbkrk) [inline packages/commutable/src/v4.js:249]: Yeah I'm definitely liking this approach, it's handling it as it gets it.

**Review by rgbkrk — APPROVED** [review]
I like this approach of including it in the commutable data structure.

We'll probably want to bring `reduceOutputs` into commutable itself too.

**Comment by rgbkrk** [comment]
> Update reduceOutput to always use createImmutableOutput.

Oh right, you already said here that we'd move reduceOutput into here. Silly me.

**Comment by rgbkrk** [comment]
By the way, my approval here means that I think the approach is :cool:.

**Comment by lgeiger** [comment]
@rgbkrk Thanks a lot for the review!

I did some further testing and unfortunately the ANSI parsing plays only a small part of the performance problem 😕 .

I recorded a performance profile of different Ansi components while executing
```python
print("This is tooooooo long :(" * 1000000)
```

- Current **master** took **9.07 s** to render
- A functional component with no ANSI escaping renders in **8.34 s**:
```js
const Ansi = props => <code className={props.className}>{props.children}</code>;
```
- A pure component with no ANSI escaping renders in **8.24 s**:
```js
class Ansi extends React.PureComponent {
  shouldComponentUpdate(nextProps) {
    return (
      this.props.children != nextProps.children ||
      this.props.className != nextProps.className
    );
  }
  render() {
    return <code className={this.props.className}>{this.props.children}</code>;
  }
}
```

The performance gain of precomputing the ANSI escape will increase when handling streaming output but this doesn't solve the rendering problem 😕 .

Please note, I only recorded the profiles once so I don't claim any statistical significance but it's clear to see that removing the ANSI parsing won't fix the rendering performance issues.

I'm happy to upload the recorded profiles if someone wants to investigate further.

**Comment by rgbkrk** [comment]
Now that I'm thinking about this some more, I think we should handle this at the level of the message processing from the stream. We'll be able to handle backpressure further up the stack and only present to React what's ready in a reasonable time.

**Comment by rgbkrk** [comment]
Did you have any more thoughts on this?

**Comment by lgeiger** [comment]
Since this doesn't solve the problem, I think we can close it for now.

**Pull request closed without merging**